sda/router.go

```diff
@@ -0,0 +1,549 @@
+// communicate.
+func NewTCPRouter(e *network.ServerIdentity, pkey abstract.Scalar) *TCPRouter {
+	return &TCPRouter{
+		Dispatcher:     NewBlockingDispatcher(),
+		workingAddress: e.First(),
+		connections:    make(map[network.ServerIdentityID]network.SecureConn),
+		host:           network.NewSecureTCPHost(pkey, e),
+		suite:          network.Suite,
+		serverIdentity: e,
+		quitProcessMsg: make(chan bool),
+		// buffered channel of 1 so Close() without
+		// Run() before does not fail
+		closing:     make(chan bool, 1),
+		networkChan: make(chan network.Packet, 1),
+	}
+}
+
+// Send sends to an ServerIdentity without wrapping the msg into a SDAMessage
+func (t *TCPRouter) Send(e *network.ServerIdentity, msg network.Body) error {
+	if msg == nil {
+		return errors.New("Can't send nil-packet")
+	}
+	t.networkLock.RLock()
+	c, ok := t.connections[e.ID]
+	t.networkLock.RUnlock()
+	if !ok {
+		var err error
+		c, err = t.Connect(e)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Lvlf4("%s sends to %s msg: %+v", t.serverIdentity.Addresses, e, msg)
+	var err error
+	err = c.Send(context.TODO(), msg)
+	if err != nil /*&& err != network.ErrClosed*/ {
+		log.Lvl2("Couldn't send to", c.ServerIdentity().First(), ":", err, "trying again")
+		c, err = t.Connect(e)
+		if err != nil {
+			return err
+		}
+		err = c.Send(context.TODO(), msg)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl5("Message sent")
+	return nil
+}
+
+// Run will start opening a tcp port and accepting connections. It is a blocking
+// call. This function returns when an error occurs on the open port or when
+// t.Stop() is called.
+func (t *TCPRouter) Run() {
+	// start processing messages
+	go func() {
+		t.networkLock.Lock()
+		t.quitProcessMsg = make(chan bool)
+		if t.processMessagesStarted {
+			// we are already listening
+			t.networkLock.Unlock()
+			return
+		}
+		t.networkLock.Unlock()
+
+		t.processMessages()
+	}()
+	// listen
+	go func() {
+		err := t.listen()
+		if err != nil {
+			log.Fatal("Error listening on", t.workingAddress, ":", err)
+		}
+	}()
+	<-t.closing
+}
+
+func (t *TCPRouter) listen() error {
+	// open port
+	log.Lvl3(t.serverIdentity.First(), "starts to listen")
+	fn := func(c network.SecureConn) {
+		log.Lvl3(t.workingAddress, "Accepted Connection from", c.Remote())
+		// register the connection once we know it's ok
+		t.registerConnection(c)
+		t.handleConn(c)
+	}
+	err := t.host.Listen(fn)
+	if err != nil {
+		log.Fatal("Error listening on", t.workingAddress, ":", err)
+		return err
+	}
+	return nil
+}
+
+// processMessages checks if it is one of the messages for us or dispatch it
+// to the corresponding instance.
+// Our messages are:
+// * SDAMessage - used to communicate between the Hosts
+// * RequestTreeID - ask the parent for a given tree
+// * SendTree - send the tree to the child
+// * RequestPeerListID - ask the parent for a given peerList
+// * SendPeerListID - send the tree to the child
```

Comment:
Old comment. Should have been removed during the pr for `refactor_processor`. Will remove it. (it belongs to `overlay` now to treat these messages)